contracts/btc-finality/src/contract.rs

```diff
@@ -505,73 +506,16 @@ pub(crate) mod tests {
-        assert_eq!(res.attributes[0], attr("action", "migrate"));
-        assert_eq!(res.attributes[1], attr("from_version", "0.1.0"));
-        assert_eq!(res.attributes[2], attr("to_version", CONTRACT_VERSION));
-
-        // Verify the version was updated
-        let version_info = cw2::get_contract_version(&deps.storage).unwrap();
-        assert_eq!(version_info.contract, CONTRACT_NAME);
-        assert_eq!(version_info.version, CONTRACT_VERSION);
-    }
-
-    #[test]
-    fn test_migrate_after_instantiate() {
-        let mut deps = mock_dependencies();
-        let msg = InstantiateMsg::default();
-        let info = message_info(&deps.api.addr_make(CREATOR), &[]);
-
-        // Instantiate the contract first
-        instantiate(deps.as_mut(), mock_env(), info, msg).unwrap();
-
-        // Verify initial version is set
-        let version_info = cw2::get_contract_version(&deps.storage).unwrap();
-        assert_eq!(version_info.contract, CONTRACT_NAME);
-        assert_eq!(version_info.version, CONTRACT_VERSION);
-
-        // Now attempt migration
-        let res = migrate(deps.as_mut(), mock_env(), crate::msg::MigrateMsg {}).unwrap();
-
-        // Check that the response contains the expected attributes
-        assert_eq!(res.attributes.len(), 3);
-        assert_eq!(res.attributes[0], attr("action", "migrate"));
-        assert_eq!(res.attributes[1], attr("from_version", CONTRACT_VERSION));
-        assert_eq!(res.attributes[2], attr("to_version", CONTRACT_VERSION));
-
-        // Verify the version remains the same
-        let version_info = cw2::get_contract_version(&deps.storage).unwrap();
-        assert_eq!(version_info.contract, CONTRACT_NAME);
-        assert_eq!(version_info.version, CONTRACT_VERSION);
-    }
-
-    #[test]
-    fn test_migrate_wrong_contract() {
-        let mut deps = mock_dependencies();
-
-        // Set a wrong contract name to simulate migration from different contract
-        cw2::set_contract_version(&mut deps.storage, "wrong-contract", "0.1.0").unwrap();
-
-        // Call migrate and expect error
-        let err = migrate(deps.as_mut(), mock_env(), crate::msg::MigrateMsg {}).unwrap_err();
-
-        // Check the error is InvalidContractName
-        match err {
-            ContractError::InvalidContractName { expected, actual } => {
-                assert_eq!(expected, CONTRACT_NAME);
-                assert_eq!(actual, "wrong-contract");
-            }
-            _ => panic!("Expected InvalidContractName error"),
-        }
+    fn test_migration_basics() {
+        MigrationTester::new(CONTRACT_NAME, CONTRACT_VERSION).test_migration_basics(
+            migrate,
+            instantiate,
+            crate::msg::MigrateMsg {},
+            InstantiateMsg::default(),
+            |err| match err {
+                ContractError::InvalidContractName { expected, actual } => Some((expected, actual)),
+                _ => None,
+            },
+        );
```

Comment:
Let's call `test_basic_migration`, `test_after_instantiate`, `test_wrong_contract` given that they are taking different parameters? Same question for other tests

Reply:
Not sure I follow your question. Are you saying you want to inline the function `test_migration_basics()` within each contract?

`test_migration_basics()` is a standard migration test suite used to contain all common test scenarios. With this unified API, we can easily extend the standard migration tests in the future.

Reply:
Yeah that was what I mean. Okay then